Add modal to compare weapons in generated builds in the optimize tab and disable individual swapping/locations

arquelion opened the pull request:
## Describe your changes
- Added `CompareWeaponModal` to `BuildDisplayItem.tsx` that functions identically to the existing `CompareArtifactModal`
- Removed the ability to swap artifacts from `CompareArtifactModal`
- Disabled the location dropdowns for both comparison modals
- Fixed a bug where the old weapon/artifact shown in the comparison modal does not reflect any changes made to the comparison build despite calcs updating in real-time
<!--- Provide a general summary of your changes -->

## Issue or discord link
- Resolves #1591 
- Resolves #1616
- https://discord.com/channels/785153694478893126/1200154094476734556/1214119792999403591<!--- link relevant issues to this PR, or provide a link to a discord message/thread -->

## Testing/validation
Comparison weapon matches optimized weapon
![image](https://github.com/frzyc/genshin-optimizer/assets/47800040/d97658c7-c535-4930-a4e2-fe1353113fc8)

Comparison weapon does not match optimized weapon
![image](https://github.com/frzyc/genshin-optimizer/assets/47800040/b4117ab4-6e04-4c84-a91a-1077fa45a07e)

Modified artifact comparison
![image](https://github.com/frzyc/genshin-optimizer/assets/47800040/38a7d327-3c9b-4855-b4f1-d71d022099db)

<!--- Add screenshots if possible -->

## Checklist before requesting a review (leave this PR as draft if any part of this list is not done.)

- [x] I have commented my code in hard-to understand areas.
- [x] I have made corresponding changes to README or wiki.
- [x] For front-end changes, I have updated the corresponding English translations.
- [x] I have run `yarn run mini-ci` locally to validate format and lint.
- [x] If I have added a new library or app, I have updated the deployment scripts to ignore changes as needed

frzyc reviewed (changes requested):
Your changes here seems to revert changes added in https://github.com/frzyc/genshin-optimizer/pull/1597, can you see if its an issue with your git history?

arquelion commented:
Looks like the revert I did after accidentally merging in master when I meant to update my own master branch undid those changes. Should be fixed now.

frzyc commented on apps/frontend/src/app/PageTeam/CharacterDisplay/Tabs/TabOptimize/Components/BuildDisplayItem.tsx:
```diff
@@ -87,8 +88,10 @@ export default function BuildDisplayItem({
   const { data } = useContext(DataContext)
-  const [newOld, setNewOld] = useState(undefined as NewOld | undefined)
-  const close = useCallback(() => setNewOld(undefined), [setNewOld])
+  const [weapNewOld, setWeapNewOld] = useState(undefined as NewOld | undefined)
```
```Typescript

// update when data is recalc'd
const weaponNewOld = useMemo(()=>({
  oldId: oldData.get(input.weapon.id).value!
  newId: data.get(input.weapon.id).value!
}),[data,oldData])

//state for showing weaponmodal
const [showWeapon, onShowWeapon, onHideWeapon] = useboolean()
```

frzyc commented on apps/frontend/src/app/PageTeam/CharacterDisplay/Tabs/TabOptimize/Components/BuildDisplayItem.tsx:
```diff
@@ -87,8 +88,10 @@ export default function BuildDisplayItem({
-  const [newOld, setNewOld] = useState(undefined as NewOld | undefined)
-  const close = useCallback(() => setNewOld(undefined), [setNewOld])
+  const [weapNewOld, setWeapNewOld] = useState(undefined as NewOld | undefined)
+  const [artNewOld, setArtNewOld] = useState(undefined as NewOld | undefined)
```
```Typescript
// update when data is recalc'd
const artNewOldBySlot = useMemo(()=>
  objKeyMap(allArtifactSlots,(slotKey)=>({
    oldId: oldData.get(input.art[slotKey].id).value ??''
    newId: data.get(input.art[slotKey].id).value ??''  
  )) as Record<SlotKey,NewOld>
,[data,oldData])

//state for showing art compare modal
const [artNewOld, setArtNewOld] = useState<NewOld|undefined>()
```

frzyc reviewed (changes requested):
I just noticed that there is a bug, where users can change the artifacts/weapon on the "compare build", and the calculation comparison on generated builds would recalculate, but the popup for comparison (new/old) of art/weapon would not be updated. 
There are two ways we can fix this. 
1. register for updates with the compared build, which can trigger UI updates,
2. Take the weapon/artifact ids from the calculated Data. 

I think the 2nd approach is more "elegant", since a recalculation data is already propagated, we can piggy back off it without adding additional listeners. I added some pseudocode to get you started.

arquelion commented:
Fixed, I've manually checked it locally but can add validation images/recordings later.

Looks like this bug was already present in top of tree `master` so I've also added the fix to the PR description

frzyc commented:
LGTM, thanks for addressing the fix as well.
